apps/total-typescript/src/ability/ability.ts

```diff
@@ -1,53 +1,98 @@
 import {Ability, AbilityBuilder, AbilityClass} from '@casl/ability'
 import {Exercise} from '../lib/exercises'
 import {SanityDocument} from '@sanity/client'
+import z from 'zod'
+import {hasValidPurchase} from '@skillrecordings/ability'
+
+const adminRoles = ['ADMIN', 'SUPERADMIN']
+
+export const UserSchema = z.object({
+  role: z.string(),
+  purchases: z.array(z.any()),
+  id: z.string(),
+  name: z.nullable(z.string().optional()),
+  email: z.string(),
+})
+
+export type User = z.infer<typeof UserSchema>
 
 type Actions = 'manage' | 'invite' | 'view'
 type Subjects =
   | 'Team'
   | 'Purchase'
   | 'Content'
+  | 'Lesson'
+  | 'Module'
   | 'Product'
   | 'Invoice'
   | 'Account'
   | 'all'
 export type AppAbility = Ability<[Actions, Subjects]>
-const AppAbility = Ability as AbilityClass<AppAbility>
+export const AppAbility = Ability as AbilityClass<AppAbility>
 
 type ViewerAbilityInput = {
-  user: any
-  subscriber: any
-  lesson: Exercise
+  user?: User
+  subscriber?: any
+  lesson?: Exercise
   module: SanityDocument
   section?: SanityDocument
+  isSolution?: boolean
 }
 
-function defineRulesForPurchases(viewerAbilityInput: ViewerAbilityInput) {
+export function defineRulesForPurchases(
+  viewerAbilityInput: ViewerAbilityInput,
+) {
   const {can, rules} = new AbilityBuilder(AppAbility)
-  const {section, module, lesson, subscriber, user} = viewerAbilityInput
+  const {
+    section,
+    module,
+    lesson,
+    subscriber,
+    user,
+    isSolution = false,
+  } = viewerAbilityInput
 
   const exercises = section ? section.exercises : module.exercises
   const isFirstLesson =
-    lesson._type === 'exercise' && lesson._id === exercises?.[0]._id
+    lesson?._type === 'exercise' && lesson._id === exercises?.[0]._id
 
-  const hasVideo = Boolean(lesson.muxPlaybackId)
+  const hasVideo = Boolean(lesson?.muxPlaybackId)
 
   if (module.moduleType === 'tutorial') {
-    if (subscriber || (!section && hasVideo)) {
+    if (subscriber || (!section && isFirstLesson && hasVideo && !isSolution)) {
       can('view', 'Content')
     }
   }
 
-  if (isFirstLesson && hasVideo) {
+  if (isFirstLesson && hasVideo && !isSolution) {
     can('view', 'Content')
   }
 
   if (module.moduleType === 'workshop') {
+    // TODO remove this once we have a better way to determine if a workshop is
+    //  available to the user (see below)
     if (user && hasVideo) {
       can('view', 'Content')
     }
+
+    // TODO a given module is associated with a product
+    //  if the user has a valid purchase of that product
+    //  they can view the content of the lesson
+    // if (hasValidPurchase(user?.purchases)) {
+    //   can('view', 'Account')
+    //   can('view', 'Content')
+    //   can('view', 'Product', {
+    //     productId: {
+    //       $in: user?.purchases?.map(
+    //         (purchase: any) => purchase.productId && purchase.status === 'Valid',
+    //       ),
+    //     },
+    //   })
+    // }
```

Comment:
ultimately we are looking for a "can view module" here based on the array of purchases (products) that would contain this module

```
can('view', 'Module', {
     moduleSlug: {
      $in: user?.purchases?.map(
        (purchase: any) => purchase.productId && purchase.product.hasModule(moduleSlug),
      ),
    },
  })
}
```

`purchase.product.hasModule(moduleSlug)` <- this doesn't exist yet